prompt2model/dataset_transformer/prompt_based.py

```diff
@@ -0,0 +1,125 @@
+"""A simple dataset transformer that uses a plan prompt and transform prompt."""
+from __future__ import annotations
+
+import asyncio
+from collections.abc import Callable
+
+import datasets
+
+from prompt2model.dataset_transformer.base import DatasetTransformer
+from prompt2model.dataset_transformer.prompt_template import (
+    construct_prompt_for_plan,
+    construct_prompt_for_transform_data,
+)
+from prompt2model.prompt_parser import PromptSpec
+from prompt2model.utils import (
+    API_ERRORS,
+    api_tools,
+    get_formatted_logger,
+    handle_api_error,
+)
+from prompt2model.utils.parse_json_responses import (
+    extract_response,
+    make_request_from_prompt,
+)
+
+logger = get_formatted_logger("DatasetTransformer")
+
+
+class PromptBasedDatasetTransformer(DatasetTransformer):
+    """Transform data based on a transform prompt."""
+
+    def __init__(
+        self,
+        plan_prompt_fn: Callable[
+            [str, list[dict], str], str
+        ] = construct_prompt_for_plan,
+        transform_prompt_fn: Callable[
+            [str, dict, str, str], str
+        ] = construct_prompt_for_transform_data,
+    ):
+        """Initialize the class."""
+        self.plan_prompt_fn = plan_prompt_fn
+        self.transform_prompt_fn = transform_prompt_fn
+        self.plan: str = ""
+
+    def canonicalize_dataset_using_samples(
+        self,
+        inputs: list[str],
+        outputs: list[str],
+    ) -> datasets.DatasetDict:
+        """Canonicalize a dataset into a suitable text-to-text format."""
```

Comment:
I don't think this function's name or docstring are faithful to what the function is actually doing. Can we provide a better name and description?